source/slang/slang-ir-use-uninitialized-values.cpp

```diff
@@ -0,0 +1,395 @@
+                return false;
+            default:
+                break;
+            }
+
+            return true;
+    }
+
+    static bool isAliasable(IRInst* inst)
+    {
+        switch (inst->getOp())
+        {
+        // These instructions generate (implicit) references to inst
+        case kIROp_FieldExtract:
+        case kIROp_FieldAddress:
+        case kIROp_GetElement:
+        case kIROp_GetElementPtr:
+            return true;
+        default:
+            break;
+        }
+
+        return false;
+    }
+
+    static bool isDifferentiableFunc(IRInst* func)
+    {
+        for (auto decor = func->getFirstDecoration(); decor; decor = decor->getNextDecoration())
+        {
+            switch (decor->getOp())
+            {
+            case kIROp_ForwardDerivativeDecoration:
+            case kIROp_ForwardDifferentiableDecoration:
+            case kIROp_BackwardDerivativeDecoration:
+            case kIROp_BackwardDifferentiableDecoration:
+            case kIROp_UserDefinedBackwardDerivativeDecoration:
+                return true;
+            default:
+                break;
+            }
+        }
+
+        return false;
+    }
+
+    static bool canIgnoreType(IRType* type)
+    {
+        if (as<IRVoidType>(type))
+            return true;
+
+        // For structs, ignore if its empty
+        if (as<IRStructType>(type))
+            return (type->getFirstChild() == nullptr);
+
+        // Nothing to initialize for a pure interface
+        if (as<IRInterfaceType>(type))
+            return true;
+
+        // For pointers, check the value type (primarily for globals)
+        if (auto ptr = as<IRPtrType>(type))
+            return canIgnoreType(ptr->getValueType());
+
+        // In the case of specializations, check returned type
+        if (auto spec = as<IRSpecialize>(type)) {
+            IRInst* base = spec->getBase();
+            IRGeneric* generic = as<IRGeneric>(base);
+            IRInst* inner = findInnerMostGenericReturnVal(generic);
+            IRType* innerType = as<IRType>(inner);
+            return canIgnoreType(innerType);
+        }
+
+        return false;
+    }
+
+    static List<IRInst*> getConcernableUsers(IRInst* inst)
+    {
+        List<IRInst*> users;
+        for (auto use = inst->firstUse; use; use = use->nextUse)
+        {
+            IRInst* user = use->getUser();
+            // Meta instructions only use the argument type
+            if (!isMetaOp(user))
+                users.add(user);
+        }
+
+        return users;
+    }
+
+    static List<IRInst*> getAliasableInstructions(IRInst* inst)
+    {
+        List<IRInst*> addresses;
+
+        auto users = getConcernableUsers(inst);
+
+        addresses.add(inst);
+        for (auto user : users)
+        {
+            if (isAliasable(user))
+            {
+                auto instAddresses = getAliasableInstructions(user);
+                addresses.addRange(instAddresses);
+            }
+        }
```

Comment:
Question: 
do we need to allocate a list here, would the following work?

```suggestion
        addresses.add(inst);
        for (auto use = inst->firstUse; use; use = use->nextUse)
        {
            IRInst* user = use->getUser();
            // Meta instructions only use the argument type
            if (isMetaOp(user) || !isAliasable(user))
                continue;
            addresses.addRange(getAliasableInstructions(user));
        }
```

Reply:
Yes, that works, thanks for the suggestion.